crates/oxc_linter/src/rules/eslint/require_await.rs

```diff
@@ -0,0 +1,124 @@
+use oxc_ast::{
+    ast::{Expression, Statement},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Default, Clone)]
+pub struct RequireAwait;
+
+fn require_await_diagnostic(span0: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("eslint(require-await): Async function has no 'await' expression.")
+        .with_labels([span0.into()])
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow async functions which have no await expression.
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// async function foo() {
+    ///   doSomething();
+    /// }
+    /// ```
+    RequireAwait,
+    pedantic,
+);
+
+impl Rule for RequireAwait {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::FunctionBody(body) = node.kind() {
+            if body.is_empty() {
+                return;
+            }
+
+            let Some(parent) = ctx.nodes().parent_kind(node.id()) else {
+                return;
+            };
+
+            match parent {
+                AstKind::Function(func) => {
+                    if func.r#async && !func.generator && !has_await(&body.statements) {
+                        ctx.diagnostic(require_await_diagnostic(func.span));
+                    }
+                }
+                AstKind::ArrowFunctionExpression(func) => {
+                    if func.r#async && !has_await(&body.statements) {
+                        ctx.diagnostic(require_await_diagnostic(func.span));
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn has_await<'a>(statements: &'a [Statement<'a>]) -> bool {
```

Comment:
This function may ignore some cases like:

```js
async function foo() {
	{	
		await doSomething()
	}
}
```

What about implementing a visit pattern for it:

```rust
struct AwaitFinder {
    found: bool,
}

impl<'a> Visit<'a> for AwaitFinder {
    fn visit_await_expression(&mut self, expr: &AwaitExpression) {
        if self.found {
            return;
        }
        self.found = true;
    }
    fn visit_for_of_statement(&mut self, stmt: &ForOfStatement) {
        if stmt.r#await {
            self.found = true;
        }
    }
	// you may need this if you visit from function body.
    fn visit_function(&mut self, func: &oxc_ast::ast::Function<'a>, flags: Option<oxc_semantic::ScopeFlags>) {
        return;
    }
	// 	you may need this if you visit from function body.
    fn visit_arrow_expression(&mut self, expr: &oxc_ast::ast::ArrowFunctionExpression<'a>) {
        return;
    }
}


 if func.r#async {
      let mut finder = AwaitFinder { found: false };
      finder.visit_function_body(body);
      if finder.found {
          return;
      }
      ctx.diagnostic(require_await_diagnostic(func.span));
  }
```

Reply:
nice! just learned something new. and good catch about that corner case